WIP: Refactoring and some new features

vortigont opened the pull request:
Приветствую, @Whilser 
я начал небольшой рефакторинг кода и исправление мелких ошибок по ходу дела. Но текущая структура скетча очень неудобна для совместной разработки и версионирования, код вперемешку с определением констант, библиотек и бинарников сложно сопровождать. Мне хотелось бы возвращать наработки в общий проект, дабы не плодить новый форк, но бекпортировать изменения в существующий скетч будет очень сложно.
Я работаю под platformio, поэтому начал менять структуру файлов и проекта в целом, выкинул комплектные библиотеки. Сборка под Arduino IDE, конечно же, сохраняется. Вам интересны эти наработки? Ваш проект ушел вперед от апстрима гайвера, поэтому создал PR в ваш реп.

Пока реализованны следущие изменения:

**BUILD:**    Adopt building with platformio
- Moved sources out of firmware
- removed bundled libs in favor of up to date versions from repo's
- moved sketch config out of .ino file
- Moved main defines to the .h file


**NTP:** obsolete NTPClient lib
    Убрана зависимость от внешней библиотеки NTPClient.
    Для синхронизации c ntp и работой с датой/временем используются фунции ядра
- поддерживается установка ntp-сервера из опций DHCP-аренды
- автонастройка правил летнего времени в зависимости от временной зоны
- временная зона устанавливается на основе определений из TZ.h
- в конфиге задается двухбуквенный код страны для автовыбора ближайшего NTP-сервера по-умолчанию
    
**Будильник:**
- будильник блокируется если время не было выставленно через ntp или потеряно соединение с WiFi
- костыли с обработкой даты/веремени переписаны на фунции ядра

**ALARM:** per-second color-gradient
Реализован посекундный градиент для будильника
- упрощен планировщик вызова
- световой градиент расчитывается посекундно в течение всего периода рассвета

**Serial:** Disabled serial output messages by default
     По-умолчанию debug сообщения и их вывод через serial port не влючается в прошивку,
     это экономит память контроллера и не тратит ресурсы на бесполезную работу
     для сборки отладочной debug-версии можно использовать отдельный таргет platformio
     (pio run -e debug) или раскомментировать _DEBUG_ флаг

**UDP:** Fix UDP parser always updates MQTT status
- Исправлена ошибка в UPD парсере, приводившая к тому что статус топики MQTT обновлялись при получении каждого UDP пакета от мобильного приложения, даже если не менялить никакие параметры

**SCHED:** Obsolete TimerMinim.h lib
переход на системный планировщик из ESP8266 core
     - минус зависимость от либы TimerMinim
     - будильник переписан в набор состояний

**SCHED:** effects/demo/UDP parser now uses scheduled events processing
            - obsolete Timer lib dependency, using core ticker for events scheduling
            - Effects processor now uses ticker scheduler and could be activated/suspended on-demand
            - Demo uses it's own ticker, easier to maintain states instead of global flags
            - touch button no longer disables demo state on power on/off, next/prev effect
            - minor button code rework to match other events scheduling
            - mqtt uses events scheduler to control effects/demo instead of global flags
    
**UDP:** parser refactoring
            - removed all long blockings in udp responder, now it tries to reply as fast all possible,
                all heavy tasks (mqtt/power/etc) are rescheduled for later processing out of UDP loop
            - fixed issue with app state inconsistencies with web/mqtt, app keepalive/configuration updates are periodic now
            - workaround app behavior with unnecessary config change requests that triggered mqtt/eeprom updates
            - dublicate code cleanup
    
**ALARM:** sync with power status
            - now alarm depends on lamp power status, if lamp is On than no need to trigger dawn function
                it is already pretty bright and funky :)
            - power switch controls alarm scheduler too
    
    TODO:
            - refactor button poller, release lockings
            - power on/off fader still blocks loop(), should be refactored to use asynch scheduler
            - get rid of locks in webserver
            - HA testing required

vortigont committed:
BUILD:    Adopt building with platformio
    Moved sources out of firmware
    removed bundled libs in favor of up to date versions from repo's
    moved sketch config out of .ino file
    Moved main defines to the .h file

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

vortigont committed:
NTP: obsolete NTPClient lib

Убрана зависимость от внешней библиотеки NTPClient.
Для синхронизации c ntp и работой с датой/временем используются фунции ядра
 - поддерживается установка ntp-сервера из опций DHCP-аренды
 - автонастройка правил летнего времени в зависимости от временной зоны
 - временная зона устанавливается на основе определений из TZ.h
 - в конфиге задается двухбуквенный код страны для автовыбора ближайшего NTP-сервера по-умолчанию

Будильник:
 - будильник блокируется если время не было выставленно через ntp или потеряно соединение с WiFi
 - костыли с обработкой даты/веремени переписаны на фунции ядра

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

vortigont committed:
ALARM: per-second color-gradient

Реализован посекундный градиент для будильника
 - упрощен планировщик вызова
 - световой градиент расчитывается посекундно в течение всего периода рассвета

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

vortigont committed:
Serial: Disabled serial output messages by default

 По-умолчанию debug сообщения и их вывод через serial port не влючается в прошивку,
 это экономит память контроллера и не тратит ресурсы на бесполезную работу
 для сборки отладочной debug-версии можно использовать отдельный таргет platformio
 (pio run -e debug) или раскомментировать _DEBUG_ флаг

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

vortigont committed:
UDP: Fix UDP parser always updates MQTT status

Исправлена ошибка в UPD парсере, приводившая к тому что статус топики MQTT обновлялись при
получении каждого UDP пакета от мобильного приложения, даже если не менялить никакие параметры

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

vortigont committed:
SCHED: Obsolete TimerMinim.h lib

Начат переход на системный планировщик из ESP8266 core
 - минус зависимость от TimerMinim
 - будильник переписан в набор состояний и один планировщик

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

Whilser commented:
Привет! В целом хорошо отношусь к совместной разработке, однако ввиду того, что большая часть людей для этого проекта использует Arduino IDE все же предлагаю использовать ее как основную платформу. Я понимаю, что Platformio намного удобней для разработки, я и сам предпочитаю его использовать, но у большинства пользователей возникают сложности даже со сборкой готового проекта по инструкции в Arduino IDE без необходимости переименования файлов и конфигурирования сборки. По хорошему данный скетч нужно разрабатывать с нуля, с технической точки зрения он выполнен очень посредственно со всеми вытекающими последствиями. Поэтому по хорошему его нужно либо полностью переписывать с нуля, либо вообще не трогать, внося только незначительные корректировки, направленные на исправление ошибок, ИМХО. Предлагаю также оставить структуру каталогов для того, чтобы новичкам было проще в этом разобраться, дополнительные функции добавлять отдельными файлами, внося в основные модули незначительные изменения при необходимости.

vortigont commented:
@Whilser Да, конечно возможность сборки под ардуино IDE остается, это само собой.
Очень хочется переписать все с нуля, согласен, код написан совсем уж "на скорую руку". Но на это, к сожалению, сейчас нет времени.
По поводу максимально сохранять статус кво - не могу согласиться. Ориентироваться на самого неискушенного пользователя это, по сути, заморозить проект. Он так и останется свалкой в которую все сложнее будет добавить что-то новое.
Для тех кому нужно "просто скачать" будет много лучше выкладывать в релизы гитхаба архив с несколькими готовыми типовыми прошивками, Это покроет 95% потребностей, даже ардуино IDE ставить не нужно. Остальные смогут собрать себе прошивку сами. Так действует, к примеру, проект [letscontrolit/ESPEasy](https://github.com/letscontrolit/ESPEasy)
А уж сваливать в одну кучу код, бинари и сторонние либы это совсем моветон.  Понятно, у гайвера несколько другой паблик, у него проекты больше нацелены на скорость выдачи и раскрутку.

vortigont committed:
MQTT: scheduler refactoring

    - now mqtt connects/updates use scheduled events instead of millis() polling
    - replaced timer lib with ticker
    - mqtt updates follows wifi connection status
    - try to handle mqtt disconnects more gracefully
    - removed useless subscriptions
    - HA discover config function is so ugly, that its better not to use it at all
    - disabled ESP restarts in case if mqtt server goes down, it might be not an esp issue at all

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

sashayohan commented:
В идеале для HA нужно с нуля писать не для Platformio или Arduino IDE, а для ESPHome

vortigont committed:
SCHED: effects/demo/UDP parser now uses scheduled events processing
        - obsolete Timer lib dependency, using core ticker for events scheduling
        - Effects processor now uses ticker scheduler and could be activated/suspended on-demand
        - Demo uses it's own ticker, easier to maintain states instead of global flags
        - touch button no longer disables demo state on power on/off, next/prev effect
        - minor button code rework to match other events scheduling
        - mqtt uses events scheduler to control effects/demo instead of global flags

UDP: parser refactoring
        - removed all long blockings in udp responder, now it tries to reply as fast all possible,
            all heavy tasks (mqtt/power/etc) are rescheduled for later processing out of UDP loop
        - fixed issue with app state inconsistencies with web/mqtt, app keepalive/configuration updates are periodic now
        - workaround app behavior with unnecessary config change requests that triggered mqtt/eeprom updates
        - dublicate code cleanup

ALARM: sync with power status
        - now alarm depends on lamp power status, if lamp is On than no need to trigger dawn function
            it is already pretty bright and funky :)
        - power switch controls alarm scheduler too

TODO:
        - refactor button poller, release lockings
        - power on/off fader still blocks loop(), should be refactored to use asynch scheduler
        - get rid of locks in webserver
        - HA testing required

    Signed-off-by: Emil Muratov <gpm@hotplug.ru>

vortigont commented:
@sashayohan Platformio и Arduino IDE это среда разработки, какая разница в чем писать, кому где удобнее. ESPHome это вообще отдельный проект - не совсем понял вас...

sashayohan commented:
> @sashayohan Platformio и Arduino IDE это среда разработки, какая разница в чем писать, кому где удобнее. ESPHome это вообще отдельный проект - не совсем понял вас...

Имел ввиду про прошивку. Что для HA удобнее открыть ESPHome и залить через него, без запуска и установки разных программ.

vortigont commented:
@sashayohan а, вы об этом... так esphome это по сути обертка-конструктор, которую тоже нужно установить и запустить :) заточенна она, в основном, на набор своих кубиков-примитивов. Для первоначальной прошивки все равно плату придется подключать по USB/Serial. Принципиальной разницы междй командами "esphome mysensor.yaml run" и "pio run -t upload" в общем-то никакой нет.
Портировать функционал гайвер-лампы под формат ESPHome можно конечно, но желающих взяться пока не видать )

vortigont committed:
LAMP: Non-blocking fader, natural brightness control
    - Implemented async fader using scheduler, no more loop locking
    - All global brigtness changes now uses fade effect via common cotrol
    - Natural brighness control now uses dim function to compensate
	for human's eye perception. Brighness 128 looks much closer
	to human's 50% of maximum. Fader effect also look much better

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

vortigont committed:
DOC: changelog update

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

vortigont committed:
Optimized HA autodiscover function  Whilser/GyverLamp#7
 - moved constants to PGM
 - using chunked publish for long messages
 - obsolete ArduinoJson lib

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

ya-erm commented:
Поддерживаю PR, т.к. исправление ошибок работы MQTT и интеграции с HA считаю достаточно важными. Сам тоже столкнулся с проблемами автоматического включения лампы, подвисания эффектов и пропуска нажатия на кнопку, и как я понимаю в форке @vortigont эти проблемы решены

vortigont commented:
> Добавьте файл TZ.h в репозиторий, чтобы не приходилось его отдельно скачивать

@ya-erm это файл из состава esp8266 core, смысла включать его в свой реп не вижу, т.к. он обновляется вместе с самим ядром при сборке. В дефолтном конфиге указана ссылка на него.

vortigont commented:
> Здесь возникает ошибка `DAWN_MAX_BRIGHT was not declared in this scope`. Скорее всего нужно прописать эту константу в `config.h.default` и `config.h` соответственно

Спасибо за отзыв, да, действительно забыл включить этот дефайн в дефолтовый конфиг. Думаю, ему там не место, вынесу в основной, вряд ли это опция которую нужно будет часто менять

vortigont commented:
> Здесь ошибка при компиляции `invalid conversion from 'const char*' to 'int'`, т.к. в TZONE например значение TZ_Europe_Moscow а это строка "MSK-3"

А приведите, пожалуйста, вывод компилятора и среду сборки? У меня норм собирается именно с #define TZONE TZ_Europe_Moscow

vortigont commented:
> Настройки MQTT, которые сейчас в gyverlamp.ino, тоже стоит в этот файл конфигурации вынести
> 
> ```
> char mqtt_password[32] = "DEVS_PASSWD";
> char mqtt_server[32] = "";
> char mqtt_user[32] = "DEVS_USER";
> char mqtt_port[10] = "1883";
> ```

Вот до рефакторинга этой части я еще не дошел. Там опции настройки MQTT есть в нескольких видах в разных местах, видимо копипасты из разных версий прошивок и я, пока еще, не разобрался какие реально нужны и используются, а какие нет.

vortigont committed:
config opts fix

Signed-off-by: Emil Muratov <gpm@hotplug.ru>

ya-erm commented:
@vortigont 
> > Здесь ошибка при компиляции `invalid conversion from 'const char*' to 'int'`, т.к. в TZONE например значение TZ_Europe_Moscow а это строка "MSK-3"
> 
> А приведите, пожалуйста, вывод компилятора и среду сборки? У меня норм собирается именно с #define TZONE TZ_Europe_Moscow

```
gyverlamp.ino:323:31: error: invalid conversion from 'const char*' to 'int' 
gyverlamp.ino:323:31: error: too few arguments to function 'void configTime(long int, int, const char*, const char*, const char*)'
```

vortigont commented:
> configTime
Похоже у вас какой-то конфликт библиотек, у этой функции нет аргументов с int
https://github.com/esp8266/Arduino/blob/b02643e7fadc5c7d618613058f70835d9c16649a/cores/esp8266/Arduino.h#L284
попробуйте почистить каталоги ./lib и ./libraries в папке проекта, может там остались файлы из основной ветки. В моем бранче там только GyverButton остался

ya-erm commented:
> configTime
> @ya-erm
> Похоже у вас какой-то конфликт библиотек, у этой функции нет аргументов с int
> https://github.com/esp8266/Arduino/blob/b02643e7fadc5c7d618613058f70835d9c16649a/cores/esp8266/Arduino.h#L284
> попробуйте почистить каталоги ./lib и ./libraries в папке проекта, может там остались файлы из основной ветки. В моем бранче там только GyverButton остался

@vortigont  Этот файл у меня функция configTime находится в файле time.cpp по пути `C:\Users\Yaroslav\.platformio\packages\framework-arduinoespressif8266\cores\esp8266\time.cpp`

vortigont commented:
@ya-erm все верно, configTime перегружен, там же в конце файла посмотрите есть оверлоад с char *.
Попробуйте собрать example, в нем такой же вызов.
https://github.com/esp8266/Arduino/blob/master/libraries/esp8266/examples/NTP-TZ-DST/NTP-TZ-DST.ino

ya-erm commented:
@vortigont понял в чём была проблема. Версия платформы Espressif 8266 в Platformio была старая, обновил до последней - всё стало ОК.
